File: reference-implementation/go/replay_buffer.go
package main

import (
	"fmt"
	"strconv"
	"sync"
)

const maxBufferSize = 10000

// requiredMethods lists JSON-RPC methods that must be replayed in full
// (not elided into gap_fill summaries) during reconnect replay.
var requiredMethods = map[string]bool{
	"notifications/apex.order.filled":              true,
	"notifications/apex.order.partially_filled":    true,
	"notifications/apex.order.rejected":            true,
	"notifications/apex.risk.kill_switch_engaged":  true,
}

// StoredEvent represents a single event stored in the replay buffer.
type StoredEvent struct {
	ID       int
	StreamID string
	Message  map[string]any
}

// ReplayBuffer is a fixed-size ring buffer that stores SSE events for replay.
type ReplayBuffer struct {
	mu                   sync.Mutex
	events               []StoredEvent
	nextID               int
	acknowledgedThroughID int
}

// NewReplayBuffer creates a new empty replay buffer.
func NewReplayBuffer() *ReplayBuffer {
	return &ReplayBuffer{
		events: make([]StoredEvent, 0, maxBufferSize),
		nextID: 1,
	}
}

// Store adds an event to the buffer and returns its string ID.
func (rb *ReplayBuffer) Store(streamID string, message map[string]any) string {
	rb.mu.Lock()
	defer rb.mu.Unlock()

	id := rb.nextID
	rb.nextID++

	rb.events = append(rb.events, StoredEvent{
		ID:       id,
		StreamID: streamID,
		Message:  message,
	})

	if len(rb.events) > maxBufferSize {
		rb.events = rb.events[1:]
	}

	return fmt.Sprintf("%d", id)
}

// Acknowledge marks events through lastEventID as acknowledged, trimming them
// from the buffer. Returns (acknowledgedThroughID, remaining buffer depth).
func (rb *ReplayBuffer) Acknowledge(lastEventID string) (string, int) {
	rb.mu.Lock()
	defer rb.mu.Unlock()

	targetID, err := strconv.Atoi(lastEventID)
	if err != nil {
		return "0", len(rb.events)
	}
	if targetID > rb.acknowledgedThroughID {
		rb.acknowledgedThroughID = targetID
	}
	cutoff := 0
	for i, e := range rb.events {
		if e.ID <= rb.acknowledgedThroughID {
			cutoff = i + 1
		} else {
			break
		}
	}
	if cutoff > 0 {
		rb.events = rb.events[cutoff:]
	}
	return strconv.Itoa(rb.acknowledgedThroughID), len(rb.events)
}

// extractMethod extracts the JSON-RPC "method" field from a stored event message.
func extractMethod(msg map[string]any) string {
	if m, ok := msg["method"]; ok {
		if s, ok := m.(string); ok {
			return s
		}
	}
	return ""
}

// ReplayAfter returns events after the given lastEventId, classifying them
// into required events (replayed in full) and gap_fill summaries for elided events.
// Returns (nil, false) if the lastEventId has been evicted from the buffer.
func (rb *ReplayBuffer) ReplayAfter(lastEventID string) ([]StoredEvent, bool) {
	rb.mu.Lock()
	defer rb.mu.Unlock()

	var cursorID int
	if _, err := fmt.Sscanf(lastEventID, "%d", &cursorID); err != nil {
		return nil, false
	}

	if len(rb.events) == 0 {
		return nil, true
	}

	oldestID := rb.events[0].ID

	// If the cursor is older than the oldest event, replay has failed
	if cursorID < oldestID {
		return nil, false
	}

	// Collect events after cursor
	var afterCursor []StoredEvent
	for _, evt := range rb.events {
		if evt.ID > cursorID {
			afterCursor = append(afterCursor, evt)
		}
	}

	if len(afterCursor) == 0 {
		return nil, true
	}

	// Classify into required / elided with gap_fill summaries
	var result []StoredEvent
	var gapFromID, gapToID, gapCount int

	flushGap := func() {
		if gapCount == 0 {
			return
		}
		gapFillID := gapToID
		gapNotif := map[string]any{
			"jsonrpc": "2.0",
			"method":  "notifications/apex.session.gap_fill",
			"params": map[string]any{
				"elided_count": gapCount,
				"from_id":      gapFromID,
				"to_id":        gapToID,
			},
		}
		result = append(result, StoredEvent{
			ID:       gapFillID,
			StreamID: "default",
			Message:  gapNotif,
		})
		gapCount = 0
	}

	for _, evt := range afterCursor {
		method := extractMethod(evt.Message)
		if requiredMethods[method] {
			// Flush any pending gap before this required event
			flushGap()
			result = append(result, evt)
		} else {
			// Accumulate into gap fill run
			if gapCount == 0 {
				gapFromID = evt.ID
			}
			gapToID = evt.ID
			gapCount++
		}
	}
	// Flush trailing gap
	flushGap()

	return result, true
}

// ReplayAfterRaw returns all events after the given lastEventId without
// gap fill classification. Used internally when raw replay is needed.
func (rb *ReplayBuffer) ReplayAfterRaw(lastEventID string) ([]StoredEvent, bool) {
	rb.mu.Lock()
	defer rb.mu.Unlock()

	var cursorID int
	if _, err := fmt.Sscanf(lastEventID, "%d", &cursorID); err != nil {
		return nil, false
	}

	if len(rb.events) == 0 {
		return nil, true
	}

	oldestID := rb.events[0].ID
	if cursorID < oldestID {
		return nil, false
	}

	var result []StoredEvent
	for _, evt := range rb.events {
		if evt.ID > cursorID {
			result = append(result, evt)
		}
	}

	return result, true
}

// LastAvailableID returns the ID of the most recently stored event, or 0 if empty.
func (rb *ReplayBuffer) LastAvailableID() int {
	rb.mu.Lock()
	defer rb.mu.Unlock()

	if len(rb.events) == 0 {
		return 0
	}
	return rb.events[len(rb.events)-1].ID
}

// OldestID returns the ID of the oldest event in the buffer, or 0 if empty.
func (rb *ReplayBuffer) OldestID() int {
	rb.mu.Lock()
	defer rb.mu.Unlock()

	if len(rb.events) == 0 {
		return 0
	}
	return rb.events[0].ID
}

File: reference-implementation/rust/src/replay_buffer.rs
use std::collections::VecDeque;
use std::sync::Mutex;

use serde_json::{json, Value};

const MAX_BUFFER_SIZE: usize = 10000;

#[derive(Debug, Clone)]
pub struct StoredEvent {
    pub id: u64,
    #[allow(dead_code)]
    pub stream_id: String,
    pub message: Value,
}

/// A replay item is either an original required event or a gap_fill marker
/// covering a run of elided (non-required) events.
#[derive(Debug, Clone)]
pub enum ReplayItem {
    Event(StoredEvent),
    GapFill {
        id: u64,
        elided_count: usize,
        from_id: u64,
        to_id: u64,
    },
}

pub enum ReplayResult {
    Items(Vec<ReplayItem>),
    Failed { oldest_available_id: Option<u64> },
}

pub struct ReplayBuffer {
    inner: Mutex<InnerBuffer>,
}

struct InnerBuffer {
    events: VecDeque<StoredEvent>,
    next_id: u64,
    acknowledged_through_id: u64,
}

/// Classify whether a JSON-RPC notification method is "required" —
/// meaning it must be replayed verbatim rather than elided into a gap fill.
fn is_required(method: &str) -> bool {
    matches!(
        method,
        "notifications/apex.order.filled"
            | "notifications/apex.order.partially_filled"
            | "notifications/apex.order.rejected"
            | "notifications/apex.risk.kill_switch_engaged"
    )
}

impl ReplayBuffer {
    pub fn new() -> Self {
        Self {
            inner: Mutex::new(InnerBuffer {
                events: VecDeque::new(),
                next_id: 1,
                acknowledged_through_id: 0,
            }),
        }
    }

    /// Store an event in the buffer. Returns the assigned event ID as a string.
    pub fn store(&self, stream_id: &str, message: Value) -> String {
        let mut inner = self.inner.lock().expect("replay buffer mutex poisoned");
        let id = inner.next_id;
        inner.next_id += 1;

        inner.events.push_back(StoredEvent {
            id,
            stream_id: stream_id.to_owned(),
            message,
        });

        if inner.events.len() > MAX_BUFFER_SIZE {
            inner.events.pop_front();
        }

        id.to_string()
    }

    /// Acknowledge all events up through `last_event_id`.
    /// Acknowledged events that are at the front of the buffer are removed.
    /// Returns `(acknowledged_through, buffer_depth)`.
    pub fn acknowledge(&self, last_event_id: &str) -> (String, usize) {
        let mut inner = self.inner.lock().expect("replay buffer mutex poisoned");
        let target_id = last_event_id.parse::<u64>().unwrap_or(0);
        inner.acknowledged_through_id = inner.acknowledged_through_id.max(target_id);
        while let Some(front) = inner.events.front() {
            if front.id <= inner.acknowledged_through_id {
                inner.events.pop_front();
            } else {
                break;
            }
        }
        (inner.acknowledged_through_id.to_string(), inner.events.len())
    }

    /// Replay events after the given last_event_id with gap-fill classification.
    ///
    /// Required events (order fills, rejections, kill switch) are sent verbatim.
    /// Consecutive runs of non-required events are collapsed into `gap_fill`
    /// markers carrying `elided_count`, `from_id`, and `to_id`.
    pub fn replay_after(&self, last_event_id: &str) -> ReplayResult {
        let cursor_id: u64 = match last_event_id.parse() {
            Ok(id) => id,
            Err(_) => return ReplayResult::Failed { oldest_available_id: None },
        };

        let mut inner = self.inner.lock().expect("replay buffer mutex poisoned");

        let oldest_id = inner.events.front().map(|e| e.id);

        // If cursor is older than oldest buffered event, replay has failed
        if let Some(oldest) = oldest_id {
            if cursor_id < oldest {
                return ReplayResult::Failed {
                    oldest_available_id: Some(oldest),
                };
            }
        }

        // Collect events after cursor first to avoid borrow conflicts
        let after_cursor: Vec<StoredEvent> = inner
            .events
            .iter()
            .filter(|e| e.id > cursor_id)
            .cloned()
            .collect();

        // Walk collected events, classifying into required vs elided
        let mut items: Vec<ReplayItem> = Vec::new();
        let mut elide_run: Option<(usize, u64, u64)> = None; // (count, from_id, to_id)

        for event in &after_cursor {
            let method = event
                .message
                .get("method")
                .and_then(|m| m.as_str())
                .unwrap_or("");

            if is_required(method) {
                // Flush any pending elide run before emitting the required event
                if let Some((count, from_id, to_id)) = elide_run.take() {
                    items.push(ReplayItem::GapFill {
                        id: to_id,
                        elided_count: count,
                        from_id,
                        to_id,
                    });
                }
                items.push(ReplayItem::Event(event.clone()));
            } else {
                // Accumulate into the current elide run
                match elide_run {
                    Some((count, from, _to)) => {
                        elide_run = Some((count + 1, from, event.id));
                    }
                    None => {
                        elide_run = Some((1, event.id, event.id));
                    }
                }
            }
        }

        // Flush trailing elide run
        if let Some((count, from_id, to_id)) = elide_run.take() {
            items.push(ReplayItem::GapFill {
                id: to_id,
                elided_count: count,
                from_id,
                to_id,
            });
        }

        ReplayResult::Items(items)
    }

    /// Get the current next_id (useful for generating replay_failed notification IDs).
    pub fn next_event_id(&self) -> u64 {
        let mut inner = self.inner.lock().expect("replay buffer mutex poisoned");
        let id = inner.next_id;
        inner.next_id += 1;
        id
    }

    /// Build a gap_fill JSON-RPC notification value.
    pub fn gap_fill_notification(_id: u64, elided_count: usize, from_id: u64, to_id: u64) -> Value {
        json!({
            "jsonrpc": "2.0",
            "method": "notifications/apex.session.gap_fill",
            "params": {
                "elided_count": elided_count,
                "from_id": from_id.to_string(),
                "to_id": to_id.to_string(),
            }
        })
    }
}
